vnext/Microsoft.ReactNative/IReactDispatcher.idl

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import "IReactPropertyBag.idl";
+
+namespace Microsoft.ReactNative {
+
+  // The delegate is used to create property value on-demand.
+  [webhosthidden]
+  delegate void ReactDispatcherCallback();
+
+  [webhosthidden]
+  interface IReactDispatcher
+  {
+    // True if dispatcher uses current thread.
+    Boolean HasThreadAccess { get; };
+
+    // Post task for the asynchronous execution.
+    void Post(ReactDispatcherCallback callback);
+  }
+
+  // Helper methods for the property bag implementation.
+  [webhosthidden]
+  static runtimeclass ReactDispatcherHelper
+  {
+    // Get or create IReactDispatcher for the current UI thread.
+    static IReactDispatcher UIThreadDispatcher{ get; };
```

Comment:
Is this exposed to consumers? From what I understood we weren't sure yet how we wanted to treat the multi-UI-thread scenario. I worry that anything we do expose would be hard to walk back.

Reply:
Oh, I think I misinterpreted the API. If this is just for getting the dispatcher of current thread that makes sense.

Reply:
This property is per thread. It does not matter what we decide in future: this code will work in any case. There is no such thing as 'perfect API'. This design works today. Tomorrow if we have new requirements we will change it.

---
In reply to: [423970645](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r423970645) [](ancestors = 423970645)

Reply:
Yes, this is about getting dispatcher of the current thread. The UIThreadDispatcherProperty can be used to set the UIThreadDispatcher in the ReactPropertyBag.

---
In reply to: [423981509](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r423981509) [](ancestors = 423981509)

Reply:
@vmoroz are external native modules expected to use this? "Tomorrow if we have new requirements we will change it." becomes painful if users start relying on this.

Reply:
Yes, the current vision is that developers of native modules can access IReactDispatcher of current UI thread using this API. This is a new API. I cannot give you 100% guarantee that this is the final one. It must tested by real usage first. I do not believe we any new API where we can provide such guarantee. I am not sure what is your concern here. Did you ever create API that you never changed?

---
In reply to: [423993539](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r423993539) [](ancestors = 423993539)

Reply:
Of course we will sometimes need to make changes. It's just much more expensive to do this for the public API of a package where we don't control consumers.

For internal code, or if we're in a monorepo setup like Office, we can just fixup old code using an API we don't like. It's much more expensive to do this for RNW. We either make sudden breaks, which has been a big cause of customer frustration, or we need to do phased deprecation, which is a but less painful, but annoying to track and maintain.

In this instance, we know we have the potential to run into some issues with the current design. It feels like we could avoid pain if we can prevent these rapid changes. E.g. expose this internally for now.

Consumers can still manually inject a queue of their own property if they're okay with it being instance global. Directing people towards that would avoid needing to deal with breaks and user frustration if we decide we can do something better.

Reply:
"In this instance, we know we have the potential to run into some issues with the current design." Could you please share what you see as an issue?
Previous design was not multi UI thread friendly. This one makes it more friendly, but we still must provide a dedicated UI thread per RNH. It is not a 'true' multi-UI concept, but we must do it today because many native components and RN itself needs it. So, as any other engineering solution we must implement some kind of 'transitional' design. But even if multi-UI will never happen this current design is still much better and more usable comparing to what we have today.

I would like to understand what type of design issue you see that makes you block this PR.

---
In reply to: [424009257](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r424009257) [](ancestors = 424009257)

Reply:
Wasn't sure how long-term multi-ui thread was. From the last discussion I was under the impression there was some active effort there, but now it sounds like it's further out. Just wanting to make sure we don't break this too quickly or box ourselves into a corner.

Signing off.

Reply:
Thank you for the sign off!
As we discussed offline before, the core issue with the multi-UI threads are the core native modules and the core RN that relies on the single UI thread. 
This change allows us to bring the UI queue to the RNH as an external entity. The RNH does not create it anymore. This design change moves us closer to the multi-UI threads if we ever do it. Plus, it enables use of the RNH for the XAML islands.

---
In reply to: [424022761](https://github.com/microsoft/react-native-windows/pull/4877#discussion_r424022761) [](ancestors = 424022761)